contractcourt/briefcase.go

```diff
@@ -1184,6 +1202,137 @@ func (b *boltArbitratorLog) checkpointContract(c ContractResolver,
 	}, func() {})
 }
 
+// InsertCanceledHTLCs inserts the given set of canceled HTLCs into the
+// database. It converts the set to a slice and persists the slice.
+//
+// NOTE: Part of the ArbitratorLog interface.
+func (b *boltArbitratorLog) InsertCanceledHTLCs(
+	canceledHTLCs fn.Set[uint64]) error {
+
+	return kvdb.Batch(b.db, func(tx kvdb.RwTx) error {
+		scopeBucket, err := tx.CreateTopLevelBucket(b.scopeKey[:])
+		if err != nil {
+			return err
+		}
+
+		var w bytes.Buffer
+		err = encodeHtlcSet(&w, canceledHTLCs)
+		if err != nil {
+			return err
+		}
+
+		return scopeBucket.Put(canceledHTLCsKey, w.Bytes())
+	})
+}
+
+// FetchCanceledHTLCs fetches the set of canceled HTLCs from the database.
+//
+// NOTE: Part of the ArbitratorLog interface.
+func (b *boltArbitratorLog) FetchCanceledHTLCs() (fn.Set[uint64], error) {
+	canceledHTLCs := fn.NewSet[uint64]()
+
+	err := kvdb.View(b.db, func(tx kvdb.RTx) error {
+		scopeBucket := tx.ReadBucket(b.scopeKey[:])
+		if scopeBucket == nil {
+			return errScopeBucketNoExist
+		}
+
+		canceledHTLCsBytes := scopeBucket.Get(canceledHTLCsKey)
+		if canceledHTLCsBytes == nil {
+			return errNoCanceledHTLCs
+		}
+
+		var r io.Reader = bytes.NewReader(canceledHTLCsBytes)
+
+		var err error
+		canceledHTLCs, err = decodeHtlcSet(r)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}, func() {
+		canceledHTLCs = fn.NewSet[uint64]()
+	})
+	if err != nil && !errors.Is(err, errNoCanceledHTLCs) &&
```

Comment:
oh i see, think we can just return `nil` in that case? there's also performance-wise implications since we are using `Batch` here, previous [benchmark](https://gist.github.com/yyforyongyu/fc82b46314d3ee7128dd8312182de785?permalink_comment_id=4463672#gistcomment-4463672).

Reply:
but this is the case where an unexpected error happened, so I think here we wanna return the error, could you elaborate more ?

Reply:
nvm I put the comment in the wrong line😂
it's a trick like this we use to avoid retries when using `Batch`,
https://github.com/lightningnetwork/lnd/blob/0dd58ee52937a545fd82108678e9a31611b7471f/channeldb/payment_control.go#L164-L187

Reply:
ahh cool, will change